lib/trino-orc/src/test/java/io/trino/orc/TestSliceDictionaryColumnReader.java

```diff
@@ -0,0 +1,135 @@
+import static io.trino.orc.metadata.CompressionKind.NONE;
+import static io.trino.orc.metadata.PostScript.HiveWriterVersion.ORIGINAL;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.nio.file.Files.readAllBytes;
+import static java.time.ZoneOffset.UTC;
+import static java.util.UUID.randomUUID;
+import static org.testng.Assert.assertTrue;
+
+public class TestSliceDictionaryColumnReader
+{
+    public static final int ROWS = 10_000_000;
+    private static final int DICTIONARY = 22;
+    private static final int MAX_STRING = 19;
+
+    @Test
+    public void testMemoryContext()
+            throws Exception
+    {
+        // create orc file
+        List<String> values = createValues();
+        File temporaryDirectory = createTempDir();
+        File orcFile = new File(temporaryDirectory, randomUUID().toString());
+        writeOrcColumnTrino(orcFile, NONE, VARCHAR, values.iterator(), new OrcWriterStats());
+
+        // prepare for read
+        OrcDataSource dataSource = new MemoryOrcDataSource(new OrcDataSourceId(orcFile.getPath()), Slices.wrappedBuffer(readAllBytes(orcFile.toPath())));
+        OrcReader orcReader = OrcReader.createOrcReader(dataSource, new OrcReaderOptions())
+                .orElseThrow(() -> new RuntimeException("File is empty"));
+        Footer footer = orcReader.getFooter();
+        List<OrcColumn> columns = orcReader.getRootColumn().getNestedColumns();
+        assertTrue(columns.size() == 1);
+        StripeReader stripeReader = new StripeReader(
+                dataSource,
+                UTC,
+                Optional.empty(),
+                footer.getTypes(),
+                ImmutableSet.copyOf(columns),
+                footer.getRowsInRowGroup(),
+                OrcPredicate.TRUE,
+                ORIGINAL,
+                new OrcMetadataReader(),
+                Optional.empty());
+        AggregatedMemoryContext memoryContext = newSimpleAggregatedMemoryContext();
+        SliceDictionaryColumnReader columnReader = new SliceDictionaryColumnReader(columns.get(0), memoryContext.newLocalMemoryContext(TestSliceDictionaryColumnReader.class.getSimpleName()), -1, false);
+
+        List<StripeInformation> stripeInformations = footer.getStripes();
+        for (StripeInformation stripeInformation : stripeInformations) {
+            Stripe stripe = stripeReader.readStripe(stripeInformation, newSimpleAggregatedMemoryContext());
+            List<RowGroup> rowGroups = stripe.getRowGroups();
+            columnReader.startStripe(stripe.getFileTimeZone(), stripe.getDictionaryStreamSources(), stripe.getColumnEncodings());
+
+            for (RowGroup rowGroup : rowGroups) {
+                columnReader.startRowGroup(rowGroup.getStreamSources());
+                columnReader.prepareNextRead(1000);
+                columnReader.readBlock();
+                // memory usage check
+                assertGreaterThanOrEqual(memoryContext.getBytes(), columnReader.getRetainedSizeInBytes());
```

Comment:
Would it be possible to actually assert on `columnReader.getRetainedSizeInBytes()` value too?

Reply:
It is hard to find a appropriate assertion on `columnReader.getRetainedSizeInBytes() `, I think. Because of member fileds of `columnReader` is private, we can not calculate retained size of `columnReader` at outside except use reflaction to count member field or make those member field as public.

Moreover, `SliceDictionaryColumnReader#getRetainedSizeInBytes` is checked effectively by set assertion on `reader.getStreamReaderRetainedSizeInBytes()` in `TestOrcReaderMemoryUsage#testVarcharTypeWithoutNulls`.